hw/ip/edn/data/edn.hjson

```diff
@@ -259,6 +259,55 @@
                 '''
         }
       ]
-    }
+    },
+    {
+      name: "ERR_CODE",
+      desc: "Hardware detection of error conditions status register",
+      swaccess: "rw",
+      hwaccess: "hwo",
+      fields: [
+        { bits: "0",
+          name: "SFIFO_RESCMD_ERR",
+          desc: '''
+                This bit is set when a reseed command fifo error has been detected.
+                Writing a zero to this bit will reset it.
+                '''
+        }
+        { bits: "1",
+          name: "SFIFO_GENCMD_ERR",
+          desc: '''
+                This is set when a generate command fifo error has been detected.
+                Writing a zero to this bit will reset it.
+                '''
+        }
+        { bits: "28",
+          name: "FIFO_WRITE_ERR",
+          desc: '''
+                This bit is set when other source bits (any bit less than bit 28) 
+                are set in this register.
+                This is the case where a full FIFO has been written.
+                Writing a zero to this bit will reset it.
```

Comment:
Right, I think I understand this now. The bits of the register are partitioned into "source bits" (currently just 0 and 1, but we could theoretically have another 26 sources). Then bit 28 is set if a full FIFO has been written. To figure out which "source" caused the write, one looks at bits 0 and 1. Is that right? If so, this first sentence isn't quite correct (it's possible for a source bit to be set but for bit 28 to be clear).

If so, it seems a little complicated. Can't we just have 3 * 2 = 6 bits in the status word, corresponding to the 3 errors times the 2 sources? I agree that the current scheme is more efficient with lots of sources and errors (total bits = num sources + num errors), but with 2 + 3 = 5, which is almost as much as 2 * 3!

Software that wants to know if something has gone wrong can compare the (32 bit) word against zero with a single read in either case. The current design also seems to have a problem if bits 0, 1, 28 and 29 are set. What command did what wrong?

Reply:
First point: you understand correctly now what the bits represent. However, the second sentence supports the first sentence. It explains that this bit is set only when the source bit fails of a certain type of error.

Second point:  This "error_code" register addition was the result of another review where this is apparently the common way in open titan to report errors. In CSRNG, there are actually 16 source bits and the same 3 type bits. This would then expand into 48 bits. I did not want to waste that much logic on this error reporting. Using this same format then for EDN, but currently there are fewer source  bits so it appears as less efficient. In this case, common format and better bit usage were chosen over the explicit bit case.

Third point: These errors are all considered very rare (silicon) error and should virtually never happen. If they do, the response should be casters up, so recovery is not an issue. Having more than one bit on is even more rare. The question is what are you going to do with this information? Bits 28 and 29 could be a design problem that ultimately was not simulated fully, so recreating both cases would have to be done in the case that both bits came on. The 30 bit will never come on, in my opinion, but was added as additional request by review. A solution to this concern would be to lock on first error. My opinion is that this is additional  logic to build and test for an error scenario that is very rare.

Reply:
Eh, ok about the signal format. I think the text still needs fixing though. I think you intend something like "this bit gets set when ...", but I originally read it as "this bit is only in the state of being set when ..." (which is false).

Probably it's best to just expand this a little. How about something more "operational", like this?
```
When a command writes to a FIFO that is already full, the hardware sets this bit and the source bit corresponding to the command type (bit zero or one).
```
You could mirror structure in the source bits, saying something like:
```
When a generate command does something illegal, the hardware will set this bit and also one of the error type bits (bit 28 or higher).
```

I also don't think you need to explain that you can write a zero to an `rw` register in order to clear the bit (it makes it sound like there's something cleverer going on than a latch that needs clearing).

Since these bits are all essentially latched error signals and they can all be cleared (or set!) by software, I think it's good to make it clear in the text that you are describing how they get set. You don't want to have to describe when they are (in the state of being) set, because that depends on what the crazy SW folks are doing :-) I think that's probably what you intended in the first place: maybe there are just some "get"s needed to disambiguate things.

Reply:
Updated the text in the err_code register. Hope the description is somewhat clearer now.